### PR TITLE
#199  W3MI, W3HE objects

### PR DESCRIPTION
**Pull request by sbcgua**
Hi Lars, I would like to contribute with the support for W3MI, W3HE. See #199, I've posted some details and questions there

### CODE REVIEW

**Comment by larshp** [comment]
cool, thanks, will take a look later today

File: zabapgit.prog.abap [hunk 1]
```diff
@@ -11476,6 +11476,344 @@ CLASS lcl_object_sfbs IMPLEMENTATION.
...
+
+    SELECT SINGLE * INTO CORRESPONDING FIELDS OF ls_key
+      FROM wwwdata
+      WHERE relid = ls_key-relid
+      AND   objid = ls_key-objid
+      AND   srtf2 = 0.
+
+    IF sy-subrc IS NOT INITIAL.
+      RETURN.
+    ENDIF.
```
Comment (larshp) [inline zabapgit.prog.abap]: looks like the above statements can be replaced with calling method lif_object~exists instead?

File: zabapgit.prog.abap [hunk 2]
```diff
@@ -11476,6 +11476,344 @@ CLASS lcl_object_sfbs IMPLEMENTATION.
...
+      ENDIF.
+      io_xml->add( iv_name = &1
+                   ig_data = ls_wwwparam-value ).
+    END-OF-DEFINITION.
+
+    w3mi_save_param 'fileextension'.
+    w3mi_save_param 'filename'.
+    w3mi_save_param 'filesize'.
+    w3mi_save_param 'mimetype'.
+    w3mi_save_param 'version'.
```
Comment (larshp) [inline zabapgit.prog.abap]: try the following instead, the add method works on simple, structured, tables or complex data structures. Similarily the READ method can be used to read the data

```abap
io_xml->add( iv_name = 'PARAM'
                      ig_data = lt_w3params ).
```

File: zabapgit.prog.abap [hunk 3]
```diff
@@ -11476,6 +11476,344 @@ CLASS lcl_object_sfbs IMPLEMENTATION.
...
+    DATA lt_w3html   TYPE STANDARD TABLE OF w3html.
+    DATA lt_w3params TYPE STANDARD TABLE OF wwwparams.
+    DATA l_xstring   TYPE xstring.
+    DATA ls_wwwparam TYPE wwwparams.
+    DATA l_size      TYPE int4.
+    DATA l_base64str TYPE string.
+    DATA lo_utility  TYPE REF TO cl_http_utility.
+
+    ls_key-relid = ms_item-obj_type+2(2).
+    ls_key-objid = ms_item-obj_name.
```
Comment (larshp) [inline zabapgit.prog.abap]: perhaps move the above 2 lines to private method? they occur in 3 different places. The method can also be reused for method exists

**Comment by larshp** [comment]
hi, see comments in https://github.com/larshp/abapGit/pull/200/files

File: zabapgit.prog.abap (hunk 2 above)
Comment (sbcgua) [inline zabapgit.prog.abap]: Awesome! I like this style :) Nice and clean code now :) Is it this New XML feature closed recently ? Does it also supports attributes in some way ?

File: zabapgit.prog.abap (hunk 3 above)
Comment (sbcgua) [inline zabapgit.prog.abap]: Agree. Done.

**Comment by sbcgua** [comment]
Commited fixes, see answers to line comments.

**Comment by sbcgua** [comment]
P.S. Somehow I lost the comment about ~exists intead of select in serialize somewhere. So just in case: it was so initially but then it appered that I need wwwdata-text so I replaced with select. It is intended :)

**Comment by larshp** [comment]
Yes, the XML stuff is new, just a few weeks old. Structures are also supported

I deleted my comment about ~exists, as my assumption was wrong

**Comment by larshp** [comment]
looks good to me, thanks! :+1:

**Pull request merged**